### PR TITLE
Update usages of AppearanceComponent.TryGetData()

### PR DESCRIPTION
**Pull request by eoineoineoin**
## About the PR 

When doing a build of Content.Client, there's a large number of warnings generated; a decent chunk of these are from using AppearanceComponent.TryGetData(), which has been marked as obsolete. This PR removes all but one of these warnings, and gets a few more along the way.

This is quite chunky, so it's split into a few CLs:
1. b2dac14c2f1949b06c2e3c985219b0a93e3a68ec Is non-controversial changes, where the systems already had access to the non-obsoleted version - just shuffling arguments around
2. a0f18e35473bb1898c237d4474d75a1f18da4969 - Contains changes where I had to make a minor change to get the data (e.g.  changing access level private->protected for some SharedApperenceSystem pointers)
3. Three slightly more complex ones - StackVisualizer, RecyclerVisualizer and DisposalUnitVisualizer needed changes to their prototypes and some movement between Content.Client/Shared/Server.

I noted a couple of things doing this:
1. I'm seeing intermittent failures from Content.IntegrationTests (but these are happening on master, too)
2. Most implementations of VisualizerSystem<T> don't call the base class in OnAppearanceChange - seems like they ought to? (I'm happy to go though and fix these, if so)
3. Some of the non-controversial systems could be converted to use VisualizerSystem - I changed this when clearly appropriate, but some systems seemed to have a few responsibilities.
4. For a clean build of Content.Client, this reduces the warning count from 370 to 282.

**Media**

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:

### CODE REVIEW

**Comment by mirrorcult** [comment]
> Most implementations of VisualizerSystem don't call the base class in OnAppearanceChange - seems like they ought to? (I'm happy to go though and fix these, if so)

this shouldnt be necessary, the base impl is just empty

File: Content.Client/Disposal/Components/DisposalUnitComponent.cs [hunk 1]
```diff
@@ -1,13 +1,45 @@
...
+        public readonly string? OverlayReady;
+
+        [DataField("overlay_full", required: true)]
+        public readonly string? OverlayFull;
+
+        [DataField("overlay_engaged", required: true)]
+        public readonly string? OverlayEngaged;
+
+        [DataField("state_flush", required: true)]
+        public readonly string? StateFlush;
```
Comment (AJCM-git) [inline Content.Client/Disposal/Components/DisposalUnitComponent.cs:33]: These should be moved to layer maps in the prototype sprite component (you can put the enum at the end of this file)

File: Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs [hunk 2]
```diff
@@ -1,104 +1,94 @@
...
+            disposalUnit.FlushAnimation.AnimationTracks.Add(flick);
             flick.LayerKey = DisposalUnitVisualLayers.Base;
-            flick.KeyFrames.Add(new AnimationTrackSpriteFlick.KeyFrame(_stateFlush, 0));
+            flick.KeyFrames.Add(new AnimationTrackSpriteFlick.KeyFrame(disposalUnit.StateFlush, 0));
 
             var sound = new AnimationTrackPlaySound();
-            _flushAnimation.AnimationTracks.Add(sound);
+            disposalUnit.FlushAnimation.AnimationTracks.Add(sound);
+
+            sound.KeyFrames.Add(new AnimationTrackPlaySound.KeyFrame(SoundSystem.GetSound(disposalUnit.FlushSound), 0));
```
Comment (AJCM-git) [inline Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs:34]: It would be better if you made this a single object initializer

File: Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs [hunk 3]
```diff
@@ -1,104 +1,94 @@
...
+
+            var entities = IoCManager.Resolve<IEntityManager>();
+            entities.EnsureComponent<AnimationPlayerComponent>(uid);
+            var appearance = entities.EnsureComponent<ClientAppearanceComponent>(uid);
 
-            sound.KeyFrames.Add(new AnimationTrackPlaySound.KeyFrame(_flushSound.GetSound(), 0));
+            if (entities.TryGetComponent(uid, out ISpriteComponent? sprite) && sprite != null)
```
Comment (AJCM-git) [inline Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs:40]: EntitySystem has proxy methods for this, you dont have to resolve, just use `EnsureComp<>(),` and `TryComp` also why are you ensuring appearance component?

File: Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs [hunk 4]
```diff
@@ -1,104 +1,94 @@
...
-                    sprite.LayerSetState(DisposalUnitVisualLayers.Base, _stateCharging);
+                    sprite.LayerSetState(DisposalUnitVisualLayers.Base, unit.StateCharging);
                     break;
                 case VisualState.Flushing:
-                    sprite.LayerSetState(DisposalUnitVisualLayers.Base, _stateAnchored);
+                    sprite.LayerSetState(DisposalUnitVisualLayers.Base, unit.StateAnchored);
 
-                    var animPlayer = entities.GetComponent<AnimationPlayerComponent>(appearance.Owner);
+                    var entities = IoCManager.Resolve<IEntityManager>();
+                    var animPlayer = entities.GetComponent<AnimationPlayerComponent>(uid);
```
Comment (AJCM-git) [inline Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs:79]: proxy methods

File: Content.Client/Recycling/RecyclerComponent.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    [ComponentReference(typeof(SharedRecyclerComponent))]
+    public sealed class RecyclerComponent : SharedRecyclerComponent
+    {
+        // The sprite layer to use when the recycler is on
+        [DataField("state_on")]
+        public readonly string SpriteLayerOn = "grinder-o1";
+
+        // The sprite layer to use when the recycler is off
+        [DataField("state_off")]
+        public readonly string SpriteLayerOff = "grinder-o0";
```
Comment (AJCM-git) [inline Content.Client/Recycling/RecyclerComponent.cs:16]: make these mapped layers in the prototype sprite component too

File: Content.Client/Recycling/RecyclerVisualizerSystem.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            base.Initialize();
+
+            SubscribeLocalEvent<RecyclerComponent, ComponentInit>(OnComponentInit);
+        }
+
+        private void OnComponentInit(EntityUid uid, RecyclerComponent recycler, ComponentInit args)
+        {
+            var entMan = IoCManager.Resolve<IEntityManager>();
+            if (!entMan.TryGetComponent(uid, out ISpriteComponent? sprite) ||
+                !entMan.TryGetComponent(uid, out AppearanceComponent? appearance))
```
Comment (AJCM-git) [inline Content.Client/Recycling/RecyclerVisualizerSystem.cs:20]: proxy methods

File: Content.Client/Recycling/RecyclerVisualizerSystem.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
+using Content.Shared.Conveyor;
+using Content.Shared.Recycling;
+using Robust.Client.GameObjects;
+
+namespace Content.Client.Recycling
+{
+    public sealed class RecyclerVisualizerSystem : VisualizerSystem<RecyclerComponent>
```
Comment (AJCM-git) [inline Content.Client/Recycling/RecyclerVisualizerSystem.cs:7]: Im think this one can be made into a `GenericVisualizer`, so take a look at them (still gonna leave the other reviews in case it isnt possible, but if you do this you can obviously disregard them)

File: Content.Shared/Stacks/StackVisualizerComponent.cs [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    ///  <seealso cref="_spriteLayers"/>
+    /// </summary>
+    [RegisterComponent]
+    public sealed class StackVisualizerComponent : Component
+    {
+        /// <summary>
+        /// Sprite layers used in stack visualizer. Sprites first in layer correspond to lower stack states
+        /// e.g. <code>_spriteLayers[0]</code> is lower stack level than <code>_spriteLayers[1]</code>.
+        /// </summary>
+        [DataField("stackLayers")] public readonly List<string> _spriteLayers = new();
```
Comment (AJCM-git) [inline Content.Shared/Stacks/StackVisualizerComponent.cs:52]: rename the fields to PascalCase

File: Content.Client/Stack/StackVisualizerSystem.cs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            base.Initialize();
+
+            SubscribeLocalEvent<StackVisualizerComponent, ComponentInit>(OnComponentInit);
+        }
+
+        private void OnComponentInit(EntityUid uid, StackVisualizerComponent stackComponent, ComponentInit args)
+        {
+            if (stackComponent._isComposite
+                && stackComponent._spriteLayers.Count > 0
+                && IoCManager.Resolve<IEntityManager>().TryGetComponent<ISpriteComponent?>(uid,
```
Comment (AJCM-git) [inline Content.Client/Stack/StackVisualizerSystem.cs:60]: proxy method

File: Content.Client/Stack/StackVisualizerSystem.cs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+            else
+            {
+                ProcessOpaqueSprites(uid, stackComponent, args.Sprite);
+            }
+        }
+
+        private void ProcessOpaqueSprites(EntityUid uid, StackVisualizerComponent stackComponent, ISpriteComponent spriteComponent)
+        {
+            // Skip processing if no actual
+            if (!AppearanceSystem.TryGetData<int>(uid, StackVisuals.Actual, out var actual)) return;
```
Comment (AJCM-git) [inline Content.Client/Stack/StackVisualizerSystem.cs:94]: dont put the return in the same line as the condition

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Just some implementation details, also, forgot to mark them in the reviews but you also need to replace the `TryGetData()` with the` TryGetData<T>() version` (the former is only there for generic visualizers)

File: Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs (hunk 3 above)
Comment (eoineoineoin) [reply Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs]: Thanks, didn't even see that property on EntitySystem. Must have missed that AppearanceComponent - leftover from the old code.

File: Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs (hunk 2 above)
Comment (eoineoineoin) [reply Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs]: It's a little awkward; AnimationsTrackSpriteFlick.KeyFrames has a protected setter, so you _have_ to go through the Add() interface. I've tidied up this code a bit, though.

File: Content.Client/Disposal/Components/DisposalUnitComponent.cs (hunk 1 above)
Comment (eoineoineoin) [reply Content.Client/Disposal/Components/DisposalUnitComponent.cs]: I _think_ I've done this the best way? This component was trying to map multiple strings (e.g., StateAnchored,StateUnAnchored,StageCharging) unto the same layer. Because there's no standard naming convention for the disposal unit RSI states (like, e.g. doors) The only nice way I could see to do that was to make the sprite have a layer for each thing the logic wanted to enable. Maybe there's a better way to do this?

Also, wasn't sure what to do with StateFlush - that string is used for initializing an animation, so maybe it can stay where it is?

When testing this, I noticed that the disposal units never got into the 'Charging' state (even on the `master` branch) so I've applied a bonus bugfix there.

File: Content.Client/Computer/ComputerVisualizer.cs [hunk 11]
```diff
@@ -43,12 +43,12 @@ public override void OnChangeData(AppearanceComponent component)
 
             if (!IoCManager.Resolve<IEntityManager>().TryGetComponent<SpriteComponent>(component.Owner, out var sprite)) return;
 
-            if (!component.TryGetData(ComputerVisuals.Powered, out bool powered))
+            if (!component.TryGetData<bool>(ComputerVisuals.Powered, out var powered))
```
Comment (AJCM-git) [inline Content.Client/Computer/ComputerVisualizer.cs:46]: you can leave this file unchanged since another PR makes this a generic visualizer and avoiding conflicts is nice

File: Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+using Robust.Client.Animations;
+using Robust.Client.GameObjects;
+using static Content.Shared.Disposal.Components.SharedDisposalUnitComponent;
+
+namespace Content.Client.Disposal.Visualizers
+{
+    public sealed class DisposalUnitVisualizerSystem : VisualizerSystem<DisposalUnitComponent>
+    {
+        [Dependency] private readonly SharedAudioSystem SoundSystem = default!;
+        private const string AnimationKey = "disposal_unit_animation";
```
Comment (AJCM-git) [inline Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs:11]: store this on the component or in a local var where its used

File: Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            SubscribeLocalEvent<DisposalUnitComponent, ComponentInit>(OnComponentInit);
+        }
+
+        private void OnComponentInit(EntityUid uid, DisposalUnitComponent disposalUnit, ComponentInit args)
+        {
+            if (!EntityManager.TryGetComponent<ISpriteComponent>(uid, out var sprite))
```
Comment (AJCM-git) [inline Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs:22]: proxy method

File: Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs (hunk 2 above)
Comment (AJCM-git) [reply Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs]: image

I mean, that shouldnt really stop you from making  it a single object initializer

File: Content.Client/Disposal/Components/DisposalUnitComponent.cs (hunk 1 above)
Comment (AJCM-git) [reply Content.Client/Disposal/Components/DisposalUnitComponent.cs]: I dont see why StateFlush is special, `LayetGetState()` returns a state id that animations can use.
Also, you can rename the rsi states if you think its making things uglier, though making them different layers is fine imo

File: Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+            if (args.Sprite == null)
+            {
+                return;
+            }
+
+            UpdateState(uid, unit, args.Sprite);
+        }
+
+        // Update visuals and tick animation
+        private void UpdateState(EntityUid uid, DisposalUnitComponent unit, ISpriteComponent sprite)
```
Comment (AJCM-git) [inline Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs:58]: require `SpriteComponent` instead of the interface here

**Review by AJCM-git — CHANGES_REQUESTED** [review]
take a look at the build fail
Also, i *think* i missed some other things that could be proxy methods or apply some of the other reviews to, if you happen to spot anything you should fix it

File: Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs (hunk 2 above)
Comment (AJCM-git) [reply Content.Client/Disposal/Visualizers/DisposalUnitVisualizerSystem.cs]: I mean, you can still make it a single object initializer
image

**Comment by metalgearsloth** [comment]
Please split this pr up or it has a high chance of sitting in review hell.

**Comment by eoineoineoin** [comment]
Yeah, agree, it's getting pretty unwieldy. I'd really like to clean up the compiler warnings, though. Does it sound reasonable if I revert all the changes except those in https://github.com/space-wizards/space-station-14/pull/13357/commits/b2dac14c2f1949b06c2e3c985219b0a93e3a68ec and https://github.com/space-wizards/space-station-14/pull/13357/commits/a0f18e35473bb1898c237d4474d75a1f18da4969 (i.e., the uncontroversial "straightforward" changes) and I'll make individual PRs for the recycler, stack, and disposal visualizers?

**Comment by eoineoineoin** [comment]
Gonna abandon this and recreate smaller PRs.
